Add from_str function in enum Resource

ManaSugi opened the pull request:
I'm sorry for pull request again because I found a mistake.

I would like to add FromStr's from_str method in the enum Resource.
If the enum Resource has the from_str method, the Resource can convert from string to the own Resource type.
This feature is useful when retrieving the RLIMIT flags of string from a setting file and converting them to the Resource type.

Nugine commented on src/errors.rs:
```diff
@@ -0,0 +1,17 @@
+
+#[derive(Error, Debug)]
+#[error("rlimit error: {0}")]
+pub struct RlimitsError(pub(crate) String);
```
`ParseResourceError` is better, for example, <https://doc.rust-lang.org/std/index.html?search=Parse>.

A parse error is not a string, it usually likes this: <https://doc.rust-lang.org/src/core/str/mod.rs.html#149-151>.

There is usually only one way to construct a parse error: `from_str`. We should not provide `From` implementations.

Nugine commented on src/lib.rs:
```diff
@@ -167,6 +171,32 @@ impl Resource {
+impl std::str::FromStr for Resource {
+    type Err = RlimitsError;
+
+    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
+        match s {
+            "RLIMIT_AS" => Ok(Resource::AS),
+            "RLIMIT_CORE" => Ok(Resource::CORE),
+            "RLIMIT_CPU" => Ok(Resource::CPU),
+            "RLIMIT_DATA" => Ok(Resource::DATA),
+            "RLIMIT_FSIZE" => Ok(Resource::FSIZE),
+            "RLIMIT_LOCKS" => Ok(Resource::LOCKS),
+            "RLIMIT_MEMLOCK" => Ok(Resource::MEMLOCK),
+            "RLIMIT_MSGQUEUE" => Ok(Resource::MSGQUEUE),
+            "RLIMIT_NICE" => Ok(Resource::NICE),
+            "RLIMIT_NOFILE" => Ok(Resource::NOFILE),
+            "RLIMIT_NPROC" => Ok(Resource::NPROC),
+            "RLIMIT_RSS" => Ok(Resource::RSS),
+            "RLIMIT_RTPRIO" => Ok(Resource::RTPRIO),
+            "RLIMIT_RTTIME" => Ok(Resource::RTTIME),
+            "RLIMIT_SIGPENDING" => Ok(Resource::SIGPENDING),
+            "RLIMIT_STACK" => Ok(Resource::STACK),
+            _ => Err(format!("invalid rlimit: {}", s).into()),
+        }
+    }
```
We should take `cfg target` into consideration.
By the way, I'm not sure cfg settings in `src/lib.rs` are correct or not.

ManaSugi commented:
I changed error handling and added `cfg target`. But, I don't know whether the error handing is correct well. Please let me know if you wonder the error handing.

Nugine commented on src/errors.rs:
```diff
@@ -0,0 +1,8 @@
+//! Error handling.
+
+use thiserror::Error;
+
+/// Library errors
+#[derive(Error, Debug)]
+#[error("rlimts parse error: {0}")]
+pub struct RlimitsResourceError(pub(crate) String);
```
1. There are only a few lines to impl `std::error::Error`. It will be better if we don't add `thiserror`and `mod errors`, but inline the code into `src/lib.rs`.

2. Error naming should follow the api guidelines, e.g. `ParseResourceError`.
  <https://rust-lang.github.io/api-guidelines/naming.html#names-use-a-consistent-word-order-c-word-order>

3. The error type here should be a sealed ZST but not a string. String errors may cause unexpected allocations.

```rust
pub struct ParseResourceError{
    _priv: ()
}
```

ManaSugi replied:
I'm sure about 1 and 2. I'll fix them. But, in case of 3, I think that users want to know the wrong `RLIMIT_***` at the failure. For example, if an user passes the wrong rlimit flag such as `from_str("RLIMIT_WRONG")`, the error should display the `RLIMIT_WRONG` string as an error information at the failure. So, I think that `ParseResourceError` should have the element of string type.

Nugine replied:
In case of 3, the caller has a `&str` as source. The caller can use [anyhow::Context](https://docs.rs/anyhow/1.0.32/anyhow/trait.Context.html) to pass the source string. The caller can also format a error message.
There is no need for `from_str` to generate a error message.

ManaSugi replied:
I understand.

Nugine commented on src/lib.rs:
```diff
@@ -149,33 +175,90 @@ pub enum Resource {
+
+            "RLIMIT_CORE" => Ok(Self::CORE),
+
+            "RLIMIT_CPU" => Ok(Self::CPU),
+
+            "RLIMIT_DATA" => Ok(Self::DATA),
+
+            "RLIMIT_FSIZE" => Ok(Self::FSIZE),
+
+            #[cfg(target_os = "linux")]
+            "RLIMIT_LOCKS" => Ok(Self::LOCKS),
+
+            #[cfg(any(target_os = "linux", target_os = "macos", target_os = "ios"))]
+            "RLIMIT_MEMLOCK" => Ok(Self::MEMLOCK),
+
+            #[cfg(target_os = "linux")]
+            "RLIMIT_MSGQUEUE" => Ok(Self::MSGQUEUE),
+
+            #[cfg(target_os = "linux")]
+            "RLIMIT_NICE" => Ok(Self::NICE),
+
+            #[cfg(any(target_os = "linux", target_os = "macos", target_os = "ios"))]
+            "RLIMIT_NOFILE" => Ok(Self::NOFILE),
+
+            #[cfg(any(target_os = "linux", target_os = "macos", target_os = "ios"))]
+            "RLIMIT_NPROC" => Ok(Self::NPROC),
+
+            #[cfg(target_os = "linux")]
+            "RLIMIT_RSS" => Ok(Self::RSS),
+
+            #[cfg(target_os = "linux")]
+            "RLIMIT_RTPRIO" => Ok(Self::RTPRIO),
+
+            #[cfg(all(target_os = "linux", target_env = "gnu"))]
+            "RLIMIT_RTTIME" => Ok(Self::RTTIME),
+
+            #[cfg(target_os = "linux")]
+            "RLIMIT_SIGPENDING" => Ok(Self::SIGPENDING),
+
+            "RLIMIT_STACK" => Ok(Self::STACK),
```
Do we need the `RLIMIT_` prefix?

ManaSugi replied:
Yes, we need the prefix because users may pass `RLIMIT_` flags to the `from_str` like arguments of `setrlimit` system call.
If we consider reading the rlimit flags from a configuration file, the flags in that file are likely to have `RLIMIT_` prefix like linux api.

Nugine replied:
Ok

Nugine commented on src/errors.rs:
```diff
@@ -0,0 +1,8 @@
+
+use thiserror::Error;
+
+/// Library errors
```
The doc is too general.
Take a look at the std library <https://doc.rust-lang.org/src/core/str/mod.rs.html#144>

ManaSugi replied:
I'm sorry. I'll fix it.

ManaSugi commented:
I updated the error handing code following review.
CI/Clippy (pull_request) failed for a missing documentation of `_priv` field of `ParseResourceError`.
However, I think that the documentation about `_priv` is not necessary.

Nugine commented:
> I updated the error handing code following your review.
> CI/Clippy (pull_request) failed for a missing documentation of `_priv` field of `ParseResourceError`.
> However, I think that the documentation about `_priv` is not necessary.

We can add `#[allow(clippy::missing_docs_in_private_items)]` on the struct or just write something for `_priv`.

ManaSugi commented:
Thank you. I fixed.
